Framework/API/src/MatrixWorkspace.cpp

```diff
@@ -83,6 +84,41 @@ MatrixWorkspace::~MatrixWorkspace() {
   }
 }
 
+Indexing::IndexTranslator MatrixWorkspace::indexTranslator() const {
+  // Workaround while IndexTranslator is not stored in MatrixWorkspace: build
+  // translator based on data in ISpectrum.
+  std::vector<specnum_t> specNums;
+  std::vector<std::vector<specnum_t>> detIDs;
+  for (size_t i = 0; i < getNumberHistograms(); ++i) {
+    const auto &spec = getSpectrum(i);
+    specNums.push_back(spec.getSpectrumNo());
+    auto set = spec.getDetectorIDs();
+    detIDs.emplace_back(set.begin(), set.end());
+  }
+  // Note: This is the local size and the local vector of spectrum numbers and
+  // detector IDs. This will not work in an MPI run.
+  Indexing::IndexTranslator t(getNumberHistograms());
+  t.setSpectrumNumbers(std::move(specNums));
+  t.setDetectorIDs(std::move(detIDs));
+  return t;
+}
+
+void MatrixWorkspace::setIndexTranslator(
+    const Indexing::IndexTranslator &translator) {
+  // Comparing the *local* size of the translator.
+  if (translator.size() != getNumberHistograms())
+    throw std::runtime_error("MatrixWorkspace::setIndexTranslator: Translator "
```

Comment:
Rather than throwing if setIndexTranslator fails. What about returning a boolean? Does it always make sense to throw if the translator cannot be set? 

I guess throwing does seem to be the policy already with the `MatrixWorkspace`.